devito/stencilkernel.py

```diff
@@ -29,11 +39,22 @@ def __init__(self, stencils, name="Kernel", compiler=None):
         stencils = stencils if isinstance(stencils, list) else [stencils]
         self.expressions = [Expression(s) for s in stencils]
 
+        # Lower all expressions to "indexed" API
+        for e in self.expressions:
+            e.indexify()
+
+        # Apply supplied substitutions
+        if subs is not None:
+            for expr in self.expressions:
+                expr.substitute(subs)
```

Comment:
Hmm, not sure about the `2n` vs. `n^2` to be honest. We could probably write it in `2n` if we make sure not to throw any partial results away, but I see no immediate need for this at this stage.